app/controllers/zendesk_webhook_controller.rb

```diff
@@ -0,0 +1,19 @@
+class ZendeskWebhookController < ApplicationController
+  skip_before_action :verify_authenticity_token, :set_visitor_id, :set_source, :set_referrer, :set_utm_state
+  before_action :authenticate
+
+  def incoming
+    head :ok if params[:zendesk_webhook].present?
+  end
+
+  private
+
+  def authenticate
+    authenticate_or_request_with_http_basic do |name, password|
+      expected_name = EnvironmentCredentials.dig(:zendesk_webhook_auth, :name)
+      expected_password = EnvironmentCredentials.dig(:zendesk_webhook_auth, :password)
+      ActiveSupport::SecurityUtils.secure_compare(name, expected_name) &&
+        ActiveSupport::SecurityUtils.secure_compare(password, expected_password)
+    end
+  end
```

Comment:
curious, did you consider using `http_basic_authenticate_with`?
https://api.rubyonrails.org/classes/ActionController/HttpAuthentication/Basic.html

(This looks fine too. Good idea to use the `secure_compare`.)

Reply:
@tdooner I actually did that at first, but it didn't work with tests :/ . `http_basic_authenticate_with` expects a username and password and is evaluated when the class is loaded, which prevented me from mocking out a fake username and password inside the test (the class is loaded before the `before` block is run). I needed it to evaluate the username & password at runtime, so went for the block-based implementation.